src/Microsoft.ML.Dnn/DnnTransform.cs

```diff
@@ -0,0 +1,1737 @@
+
+                            var editor = VBufferEditor.Create(ref dst, (int)tensorSize);
+
+                            DnnUtils.FetchData<T>(tensor.Data<T>(), editor.Values);
+                            dst = editor.Commit();
+                        };
+                        return valuegetter;
+                    }
+                }
+            }
+
+            private void UpdateCacheIfNeeded(long position, ITensorValueGetter[] srcTensorGetters, string[] activeOutputColNames, OutputCache outputCache)
+            {
+                if (outputCache.Position != position)
+                {
+                    bool addToTheBag = false;
+                    if (!_runners.TryTake(out Runner runner))
+                    {
+                        runner = _parent.Runner.CloneRunner();
+                        addToTheBag = true;
+                    }
+                    for (int i = 0; i < _inputColIndices.Length; i++)
+                    {
+                        var inputName = _parent.Inputs[i];
+                        runner.AddInput(i, srcTensorGetters[i].GetTensor());
+                    }
+
+                    var tensors = runner.Run();
+                    Contracts.Assert(tensors.Length > 0);
+
+                    for (int j = 0; j < activeOutputColNames.Length; j++)
+                        outputCache.Outputs[activeOutputColNames[j]] = tensors[j];
+
+                    outputCache.Position = position;
+
+                    if (addToTheBag)
+                        _runners.Add(runner);
+                }
+            }
+
+            private protected override Func<int, bool> GetDependenciesCore(Func<int, bool> activeOutput)
+            {
+                return col => Enumerable.Range(0, _parent.Outputs.Length).Any(i => activeOutput(i)) && _inputColIndices.Any(i => i == col);
+            }
+
+            protected override DataViewSchema.DetachedColumn[] GetOutputColumnsCore()
+            {
+                var info = new DataViewSchema.DetachedColumn[_parent.Outputs.Length];
+                for (int i = 0; i < _parent.Outputs.Length; i++)
+                    info[i] = new DataViewSchema.DetachedColumn(_parent.Outputs[i], _parent.OutputTypes[i], null);
+                return info;
+            }
+        }
+
+        private interface ITensorValueGetter
+        {
+            Tensor GetTensor();
+
+            void BufferTrainingData();
+
+            Tensor GetBufferedBatchTensor();
+        }
+
+        private class TensorValueGetter<T> : ITensorValueGetter
+        {
+            private readonly ValueGetter<T> _srcgetter;
+            private readonly T[] _bufferedData;
+            private readonly Int64[] _bufferedDataLong;
+            private readonly TensorShape _tfShape;
+            private int _position;
+            private readonly bool _keyType;
+
+            public TensorValueGetter(
```

Comment:
If we're thinking about having Torch as another possibility for the ImageClassifier trainer we probably don't want to have the word TensorFlow hard-coded for the Label as `TensorFlowLabel` if the `Options `class will be generic for any DNN model, but maybe something generic like `DnnModelLabel`.

On the other hand, if the `Options `class will be specific for TensorFlow because the options/configurations are very different in TF compared to Torch, then we should probably call it something like `TensorFlowModelOptions` class and have another class in the future like `TorchModelOptions`.

Reply:
@CESARDELATORRE These are internal methods not public facing APIs.

Reply:
Even if those methods are internal, it'd be better if the code is consistent, I think.
The code is also open source, so anyone can research it.

Reply:
These are all internal classes and right now backend is TF specific. I agree it will be nice to not have TF in the name, may be I can make that change before next release. 

---
In reply to: [310278510](https://github.com/dotnet/machinelearning/pull/4057#discussion_r310278510) [](ancestors = 310278510)

Reply:
Right now everything is using TF so its probably ok to leave there but before we GA we should change the names to be generic. I will suspect this code will change significantly in the next month as we add more scenarios.

---
In reply to: [310331337](https://github.com/dotnet/machinelearning/pull/4057#discussion_r310331337) [](ancestors = 310331337)